dragonfly/block/farmland.go

```diff
@@ -0,0 +1,90 @@
+package block
+
+import (
+	"github.com/df-mc/dragonfly/dragonfly/item"
+	"github.com/df-mc/dragonfly/dragonfly/world"
+	"math/rand"
+)
+
+// Farmland is a block that crops are grown on. Farmland is created by interacting with a grass or dirt block using a
+// hoe. Farmland can be hydrated by nearby water, with no hydration resulting in it turning into a dirt block.
+type Farmland struct {
+	noNBT
+	tilledGrass
+
+	// Hydration is how much moisture the farmland block has. Hydration starts at 0 & caps at 7. During a random tick
+	// update, if there is water within 4 blocks from the farmland block, hydration is set to 7. Otherwise, it
+	// decrements until it turns into dirt.
+	Hydration int
+}
+
+//TODO: Add crop trampling
+
+// NeighbourUpdateTick ...
+func (f Farmland) NeighbourUpdateTick(pos, _ world.BlockPos, w *world.World) {
+	if solid := w.Block(pos.Side(world.FaceUp)).Model().FaceSolid(pos.Side(world.FaceUp), world.FaceDown, w); solid {
+		w.SetBlock(pos, Dirt{})
+	}
+}
+
+// RandomTick ...
+func (f Farmland) RandomTick(pos world.BlockPos, w *world.World, _ *rand.Rand) {
+	if !f.hydrated(pos, w) {
+		if f.Hydration > 0 {
+			f.Hydration--
+			w.PlaceBlock(pos, f)
+		} else {
+			blockAbove := w.Block(pos.Side(world.FaceUp))
+			if _, cropAbove := blockAbove.(Crop); !cropAbove {
+				w.PlaceBlock(pos, Dirt{})
+			}
+		}
+	} else {
+		f.Hydration = 7
+		w.PlaceBlock(pos, f)
+	}
+}
+
+// hydrated checks for water within 4 blocks in each direction from the farmland.
+func (f Farmland) hydrated(pos world.BlockPos, w *world.World) bool {
+	for y := 0; y <= 1; y++ {
+		for x := -4; x <= 4; x++ {
+			for z := -4; z <= 4; z++ {
+				if liquid, ok := w.Liquid(world.BlockPos{pos.X() + x, pos.Y() + y, pos.Z() + z}); ok {
```

Comment:
Another quite hot code path, might be better to grab the pos.X(), pos.Y() and pos.Z() outside of the loops at the top of the function.